java/daikon/dcomp/DCInstrument24.java

```diff
@@ -0,0 +1,4916 @@
+      il.add(ConstantInstruction.ofIntrinsic(Opcode.ACONST_NULL));
+
+      // Add the DCompMarker to the arg types list.
+      List<ClassDesc> new_arg_types = new ArrayList<>(Arrays.asList(paramTypes));
+      new_arg_types.add(dcomp_marker);
+
+      NameAndTypeEntry nte =
+          poolBuilder.nameAndTypeEntry(methodName, MethodTypeDesc.of(returnType, new_arg_types));
+      il.add(InvokeInstruction.of(invoke.opcode(), invoke.owner(), nte, invoke.isInterface()));
+      return il;
+
+    } else { // not instrumented, discard the tags before making the call
+      return cleanInvokeTagStack(invoke, classname, returnType, paramTypes);
+    }
+  }
+
+  /**
+   * Process an invoke instruction that calls a non-instrumented method. We need to clean up the tag
+   * stack.
+   *
+   * @param invoke a method invocation bytecode instruction
+   * @param classname target class of the invoke
+   * @param returnType return type of method
+   * @param paramTypes parameter types of target method
+   * @return instructions to replace the given instruction
+   */
+  private List<CodeElement> cleanInvokeTagStack(
+      Instruction invoke, String classname, ClassDesc returnType, ClassDesc[] paramTypes) {
+    List<CodeElement> il = new ArrayList<>();
+
+    // JUnit test classes are a bit strange.  They are marked as not being callee_instrumented
+    // because they do not have the dcomp_marker added to the argument list, but
+    // they actually contain instrumentation code.  So we do not want to discard
+    // the primitive tags prior to the call.
+    if (!junitTestClasses.contains(classname)) {
+      il = discard_primitive_tags(paramTypes);
+    }
+
+    // Add a tag for the return type if it is primitive.
+    if (is_primitive(returnType)) {
+      if (debugHandleInvoke) {
+        System.out.printf("push tag for return type of %s%n", returnType);
+      }
+      il.add(dcr_call("push_const", CD_void, noArgsCD));
+    }
+    il.add(invoke);
+    return il;
+  }
+
+  /**
+   * Returns instructions that will discard any primitive tags corresponding to the specified
+   * parameters. Returns an empty instruction list if there are no primitive parameters to discard.
+   *
+   * @param paramTypes parameter types of target method
+   * @return an instruction list that discards primitive tags from DCRuntime's per-thread
+   *     comparability data stack
+   */
+  private List<CodeElement> discard_primitive_tags(ClassDesc[] paramTypes) {
+    List<CodeElement> il = new ArrayList<>();
+
+    int primitive_cnt = 0;
+    for (ClassDesc paramType : paramTypes) {
+      if (paramType.isPrimitive()) {
+        primitive_cnt++;
+      }
+    }
+    if (primitive_cnt > 0) {
+      il.addAll(discard_tag_code(null, primitive_cnt));
+    }
+    return il;
+  }
+
+  /**
+   * Returns true if the invoke target is instrumented.
+   *
+   * @param invoke instruction whose target is to be checked
+   * @param mgen host method of invoke
```

Comment:
I'm not sure what "host method" means.  Is it the method that defines the caller?  It it the method that contains the call?